content/src/content/jcr_root/apps/core/wcm/sandbox/commons/ui/components/childreneditor/clientlibs/js/childreneditor.js

```diff
@@ -0,0 +1,273 @@
+/*******************************************************************************
+ * Copyright 2018 Adobe Systems Incorporated
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ ******************************************************************************/
+(function($, ns, channel, window) {
+    "use strict";
+
+    var NS = ".cmp-childreneditor";
+    var NN_PREFIX = "item_";
+    var PN_TITLE = "jcr:title";
+    var PN_RESOURCE_TYPE = "sling:resourceType";
+    var POST_SUFFIX = ".children.html";
+
+    var selectors = {
+        self: "[data-cmp-is='childreneditor']",
+        add: "[data-cmp-hook-childreneditor='add']",
+        insertComponentDialog: {
+            self: "coral-dialog.InsertComponentDialog",
+            selectList: "coral-selectlist"
+        },
+        item: {
+            icon: "[data-cmp-hook-childreneditor='itemIcon']",
+            input: "[data-cmp-hook-childreneditor='itemInput']",
+            hiddenInput: "[data-cmp-hook-childreneditor='itemHiddenInput']"
+        }
+    };
+
+    /**
+     * @typedef {Object} ChildrenEditorConfig Represents a Children Editor configuration object
+     * @property {HTMLElement} el The HTMLElement representing this Children Editor
+     */
+
+    /**
+     * Children Editor
+     *
+     * @class ChildrenEditor
+     * @classdesc A Children Editor is a dialog component based on a multifield that allows editing (adding, removing, renaming, re-ordering)
+     * the child items of panel container components.
+     * @param {ChildrenEditorConfig} config The Children Editor configuration object
+     */
+    var ChildrenEditor = function(config) {
```

Comment:
Should we declare `ChildrenEditor` as a Class?

Reply:
We do already declare it a class, or do you mean we shouldn't do so?

Reply:
ah right. I believe support in IE11 is lacking, which is still a technical requirement on AEM 6.3, 6.4 https://helpx.adobe.com/experience-manager/6-4/sites/deploying/using/technical-requirements.html

Reply:
good catch!

Reply:
Should we define `ChildrenEditor` the same way as `PanelSelector`?
`CQ.CoreComponents.PanelSelector = ns.util.createClass({`

Reply:
It was originally like that, but I had to remove it. We can't rely on editor code in the dialog as it can be opened in a new window - where no editor JS namespaces are available.

We had to add a workaround in the childreneditor to remove the "add" button in that scenario as the insert component dialog isn't available.

Reply:
I also wonder if we should make `PanelSelector` a private class within the `panelSelect.js` action file, we don't really want it to be used outside of this use case.

Reply:
ok, thx for clarifying above.